agent/grpc/client.go

```diff
@@ -102,17 +145,46 @@ func (c *ClientConnPool) dial(datacenter string, serverType string) (*grpc.Clien
 
 // newDialer returns a gRPC dialer function that conditionally wraps the connection
 // with TLS based on the Server.useTLS value.
-func newDialer(servers ServerLocator, wrapper TLSWrapper, useTLSForDC func(dc string) bool) func(context.Context, string) (net.Conn, error) {
-	return func(ctx context.Context, addr string) (net.Conn, error) {
-		d := net.Dialer{}
-		conn, err := d.DialContext(ctx, "tcp", addr)
+func newDialer(
+	servers ServerLocator,
+	srcAddr *net.TCPAddr, // SrcAddr is the source address for outgoing connections.
+	gwResolverDep *gatewayResolverDep,
+	wrapper TLSWrapper,
+	alpnWrapper ALPNWrapper,
+	useTLSForDC func(dc string) bool,
+	// dialingFromServer should be set to true if this connection pool is
+	// configured in a server instead of a client.
+	dialingFromServer bool,
+	dialingFromDatacenter string,
+) func(context.Context, string) (net.Conn, error) {
+	return func(ctx context.Context, globalAddr string) (net.Conn, error) {
+		server, err := servers.ServerForGlobalAddr(globalAddr)
 		if err != nil {
 			return nil, err
 		}
 
-		server, err := servers.ServerForAddr(addr)
+		if dialingFromServer &&
+			gwResolverDep.GatewayResolver != nil &&
+			alpnWrapper != nil &&
+			server.Datacenter != dialingFromDatacenter {
+			// NOTE: TLS is required on this branch.
+			conn, _, err := pool.DialRPCViaMeshGateway(
+				ctx,
+				server.Datacenter,
+				server.ShortName,
+				srcAddr,
+				tlsutil.ALPNWrapper(alpnWrapper),
+				pool.ALPN_RPCGRPC,
+				dialingFromServer,
+				gwResolverDep.GatewayResolver,
+			)
+			return conn, err
+		}
+
+		// TODO(rb): should this get the standard rpc dial timeout of 10 seconds?
```

Comment:
Probably? it doesn't seem like we're using timeouts at all for gRPC since we're not passing `WithTimeout` into the `grpc.Dial` nor are we using `DialContext` and passing a context with a deadline:
https://pkg.go.dev/google.golang.org/grpc#WithTimeout